1.1 API

gkellogg opened the pull request:
* **BREAKING**: Versions of Python before 3.4 are no longer supported.
* Update conformance docs.
* Add all keywords and update options.
* Default processingMode to json-ld-1.1.
* Implement logic for marking tests as pending, so that it will fail if a pending test passes.
* Consolidate documentLoader option and defaults into a load_document method to also handle JSON (eventually HTML) parsing.
* Add support for rel=alternate for non-JSON-LD docs.
* Context processing:
  * Support `@propagate` in context processing and propagate option.
  * Support for `@import`. (Some issues confusing recursion errors for invalid contexts).
  * Make `override_protected` and `propagate` optional arguments to `_create_term_definition` and `_process_context` instead of using option argument.
  * Improve management of previous contexts.
  * Imported contexts must resolve to an object.
  * Do remote context processing from within _process_contexts, as logic is too complicated for pre-loading. Removes _find_context_urls and _retrieve_context_urls.
* Create Term Definition:
  * Allow `@type` as a term under certain circumstances.
  * Reject and warn on keyword-like terms.
  * Support protected term definitions.
  * Look for keyword patterns and warn/return.
  * Look for terms that are compact IRIs that don't expand to the same thing.
  * Basic support for `@json` and `@none` as values of `@type`.
  * If `@container` includes `@type`, `@type` must be `@id` or `@vocab`.
  * Support `@index` and `@direction`.
  * Corner-case checking for `@prefix`.
  * Validate scoped contexts even if not used.
  * Support relative vocabulary IRIs.
  * Fix check that term has the form of an IRI.
  * Delay adding mapping to end of _create_term_definition.
  * If a scoped context is null, wrap it in an array so it doesn't seem to be undefined.
* IRI Expansion:
  * Find keyword patterns.
  * Don't treat terms starting with a colon as IRIs.
  * Only return a resulting IRI if it is absolute.
  * Fix _is_absolute_iri to use a reasonable regular expression and some other _expand_iri issues.
  * Fix to detecting relative IRIs.
  * Fix special case where relative path should not have a leading '/'
* Compaction:
  * Compact `@direction`.
  * Compact `@type`: `@none`.
  * Compact `@included`.
  * Honor `@container`: `@set` on `@type`.
  * Lists of Lists.
  * Improve handling of scoped contexts and propagate.
  * Improve map compaction, including indexed properties.
  * Catch Absolute IRI confused with prefix.
* Expansion:
  * Updates to expansion algorithm.
  * _expand_value adds `@direction` from term definition.
  * JSON Literals.
  * support `@direction` when expanding.
  * support lists of lists.
  * support property indexes.
  * improve graph container expansion.
  * Order types when applying scoped contexts.
  * Use type_scoped_ctx when expanding values of `@type`.
  * Use propagate and override_protected properly when creating expansion contexts.
* Flattening:
  * Rewrite _create_node_map based on 1.1 algorithm.
  * Flatten `@included`.
  * Flatten lists of lists.
  * Update merge_node_maps for `@type`.
* ToRdf:
  * Support for I18N direction.
  * Support for Lists of Lists.
  * Partial support for JSON canonicalization of JSON literals.
    * Includes local copy of JCS library, but doesn't load.
  * Lists of Lists.
  * Text Direction 'i18n-datatype'.

gkellogg committed:
Update conformance docs.

gkellogg committed:
Add all keywords and update options.

gkellogg committed:
Default processingMode to json-ld-1.1.

gkellogg committed:
Implement logic for marking tests as pending, so that it will fail if a pending test passes.

gkellogg committed:
Support `@propagate` in context processing and propagate option.

gkellogg committed:
Allow `@type` as a term under certain circumstances.
Reject and warn on keyword-like terms.
Support protected term definitions.

gkellogg committed:
Fix some pending errors.

gkellogg committed:
More work on term definitions:
* Look for keyword patterns and warn/return.
* Look for terms that are compact IRIs that don't expand to the same thing.
* Basic support for `@json` and `@none` as values of `@type`.
* If `@container` includes `@type`, `@type` must be `@id` or `@vocab`.
* Support `@index` and `@direction`.
* More corner-case checking for `@prefix`.

gkellogg committed:
* Find keyword patterns earlier for `@reverse`.
* Find keyword patterns when expanding IRIs.
* Don't treat terms starting with a colon as IRIs when expanding.
* Only return a resulting IRI if it is absolute.

gkellogg committed:
Use Python 2 syntax for merging Maps.

gkellogg committed:
Validate scoped contexts even if not used.

gkellogg committed:
* Support for `@import`. (Some issues confusing recursion errors for invalid contexts).

gkellogg committed:
_expand_value adds `@direction` from term definition.

gkellogg committed:
Make `override_protected` and `propagate` optional arguments to `_create_term_definition` and `_process_context` instead of using option argument.

gkellogg committed:
Updates to expansion algorithm.

davidlehn commented:
There's also a CHANGELOG.md file. I'm not sure what level of detail should eventually go in there for such a large batch of changes.  At least something like "Support JSON-LD 1.1.". ;-)  Probably also add any important API changes or similar things that 1.0 users would need to know when upgrading.

gkellogg commented:
I figured that I would use the PR summary for the changelig when I’m closer to being ready for a merge. 

It’s going pretty well, and now need to whittle away at the remaining pending expansion tests, as I’ve made a pass through all the relevant algorithms.

gkellogg committed:
More work on expanding text direction.

gkellogg committed:
Work on expanding property indexes.

gkellogg committed:
Expand lists of lists.

gkellogg committed:
Improve graph container expansion.

gkellogg committed:
Relative vocabulary IRIs.

gkellogg committed:
Fix check that term has the form of an IRI and delay adding mapping to end of _create_term_definition.

gkellogg committed:
Fix _is_absolute_iri to use a reasonable regular expression and some other _expand_iri issues.

gkellogg committed:
Fix to detecting relative IRIs.

gkellogg committed:
Support expanding JSON literals.

gkellogg committed:
Update merge_node_maps for `@type`.

gkellogg committed:
Compare normalized RDF results in tests.

gkellogg committed:
Test JSON results unordered.

gkellogg committed:
Improve management of previous contexts.

gkellogg committed:
Order types when applying scoped contexts.

gkellogg committed:
Use type_scoped_ctx when expanding values of `@type`.

gkellogg committed:
If a scoped context is null, wrap it in an array so it doesn't seem to be undefined.
Use propagate and override_protected properly when creating expansion contexts.

gkellogg committed:
Remaining protected test cases.

gkellogg committed:
Imported contexts must resolve to an object.

gkellogg committed:
Update changelog.

gkellogg commented:
This does everything for expansion other than HTML and some remote-doc tests which have probably never worked properly.

Some things I can't resolve and will require a change in approach:

* tests so01, so03, and so04 should get an "invalid context entry" but instead get "recursive context inclusion". This is because of pre-fetching. It seems to me that jsonld.js must have the same issue, but seems to pass these tests.
* test js21 references an unused context, but it fails when loading, similar issue.

gkellogg committed:
Mark framing/0016 pending as an emergent issue.

davidlehn commented on README.rst:
```diff
@@ -42,39 +42,15 @@ Conformance
-  2014-01-16, and any `errata`_
-- `JSON-LD 1.0 Framing <JSON-LD 1.0 Framing_>`_,
-  Unofficial Draft,
-  2012-08-30
```
The 1.1 support update should still work for most 1.0 data right?  There are some exceptions but I didn't think enough to remove this text.  I imagine the old framing and 1.1 CG draft support is more questionable at this point.

gkellogg replied:
The 1.1 specs are compatible with the 1.0 steps; it's redundant to list all previous versions of specifications supported. This includes Framing 1.1 being compatible with the unofficial 1.0 draft.

davidlehn commented on lib/pyld/jsonld.py:
```diff
@@ -4697,9 +5242,21 @@ def _find_context_urls(self, input_, urls, replace, base):
                         urls[v] = False
                 elif _is_object(v):
-                    # look for soped context
+                    # look for sooped context
```
sooped context... nomnom.
```suggestion
                    # look for scoped context
```

davidlehn commented:
I do wish we had split that file up similar to jsonld.js.  Kind of a monster to look at now!

gkellogg committed:
Update lib/pyld/jsonld.py

Co-Authored-By: David I. Lehn <dil@lehn.org>

gkellogg committed:
* ToRdf:
  * Support for I18N direction.
  * Support for Lists of Lists.

gkellogg committed:
Consolodate documentLoader option and defaults into a load_document method to also handle JSON (eventually HTML) parsing.

gkellogg committed:
Do remote context processing from within _process_contexts, as logic is too complicated for pre-loading. Removes _find_context_urls and _retrieve_context_urls.

gkellogg committed:
Fix special case where relative path should not have a leading '/'

gkellogg committed:
Flatten `@included`

gkellogg committed:
Partial support for JSON canonicalicalization of JSON literals.
 * Includes local copy of JCS library, but doesn't load.

gkellogg commented:
@davidlehn I hit a wall with both remote and local installation of the JCS module. I've included it here, but it's not loading, and I really don't understand how to manage this environment. Any help you can provide would be most useful.

I commented out the module include and the use within `_object_to_rdf`, just using `json.dumps` for now.

davidlehn committed:
Fix canonicalize loading.

- Remove unrelated 'canonicalize' package from requirements.txt.
- Fix package namespaces.
- Incorrect fix for value type.

davidlehn commented:
I got it to try and do remote install, but the lack of a setup.py seems to be a dealbreaker.
```sh
pip install "git+https://github.com/cyberphone/json-canonicalization@571690e6c9af5f614a29bc0e6034fafec93f34c8#egg=org.webpki.json.Canonicalize&subdirectory=python3"
```
Not sure if that's correct (I'm guessing), but fails due to lack of setup.py.  Would be a `[-e]` line in requirements.txt if it could be installed.  I think we'll have to deal with a copy for now.

gkellogg committed:
To RDF JSON canonicalization.

gkellogg committed:
Fix aiohttp loader, and raise 'loading document failed' if nothing returned.

gkellogg committed:
Remove support for Python < 3.4 and pypy(2).

davidlehn committed:
Remove more Python 2 support.

gkellogg committed:
Run some remote-doc tests locally, as there is no realy support for HTTP headers from the test location.

gkellogg committed:
Add support for rel=alternate for non-JSON-LD docs.

gkellogg committed:
Support ToRdf syntax tests.

gkellogg committed:
* Rewrite _create_node_map based on 1.1 algorithm.
  * Flatten lists of lists.

gkellogg committed:
Support Lists of Lists in to_rdf.

gkellogg committed:
Text Direction 'i18n-datatype'.

gkellogg committed:
Return string from document loaders and decode in load_document.
Include ContentType of returned content.

gkellogg committed:
Use lxml.html to load HTML and parse in load_html.
For HTML, the API base option can be updated from base element.

gkellogg committed:
Change the default for _expandAllScripts_ to `true` for flatten and toRdf.

gkellogg committed:
Allow 3.4 to fail.

(cherry picked from commit 1f337684523615bb75a962afd2383652d235b471)

gkellogg committed:
* Compact `@direction`
  * Compact `@type`: `@none`

gkellogg committed:
* Honor `@container`: `@set` on `@type`.

gkellogg committed:
Compact Lists of Lists.

gkellogg committed:
Improve handling of scoped contexts and propagate on compaction.

gkellogg committed:
Compact `@included`.

gkellogg committed:
Improve map compaction, including indexed properties.

gkellogg committed:
Catch Absolute IRI confused with prefix.

gkellogg committed:
Make sure '_prefix' is defined on all term definitions.

(cherry picked from commit fcce5b7507b8f324ffa3b4480984d7658b557f4d)

gkellogg committed:
Framing updates for 1.1.

* Change default for `requireAll` from True to False.
* Change default for `embed` from `@last` to `@once`.
* Add defaults for `omitGraph` and `pruneBlankNodeIdentifiers`
  based on processing mode.
* Change `_remove_preserve` to `_cleanup_preserve` which happens before
  compaction.
* Add `_cleanup_null` which happens after compaction.

gkellogg committed:
Update frame matching and support `@included`.

* Update frame matching to 1.1 spec.
* Support `@included`.

gkellogg committed:
More finessing of IRI resolution for expand#0129.

gkellogg committed:
Add doap:release to EARL report output.

Note, the version shouldn't be hard coded.

gkellogg commented on .travis.yml:
```diff
@@ -2,8 +2,6 @@ dist: xenial
 python:
-  - "2.7"
-  - "pypy"
   - "3.4"
```
Probably want to just require 3.5 or better, as I've introduced some code and programming patterns that require it. (other stuff should be cleaned up).

gkellogg commented on README.rst:
```diff
@@ -83,7 +59,7 @@ yet supported.
 ------------
 
-- Python_ (2.7 or later)
+- Python_ (3.4 or later)
```
3.5 or better, I think.

gkellogg commented on tests/runtests.py:
```diff
@@ -241,6 +233,12 @@ def setUp(self):
             if re.match(regex, data.get('@id', data.get('id', ''))):
                 self.skipTest('Test with id regex %s' % regex)
 
+        # mark tests as pending, meaning that they are expected to fail
```
Better than skipping tests that aren't ready, as it will fail when it starts passing.

gkellogg commented on tests/runtests.py:
```diff
@@ -485,6 +567,13 @@ def __init__(self):
             'doap:description': 'A JSON-LD processor for Python',
             'doap:programming-language': 'Python',
+            'doap:release': {
+                'doap:revision': '2.0.0', # FIXME: this should come from a file
```
This should come out of a file, obviously.

gkellogg commented:
A comprehensive EARL report requires running both /api and /framing test suites and merging the result using some external tooling. Could just submit as N-Triples or Turtle by converting JSON-LD output.

gkellogg committed:
Use __about__.py to get version for EARL report.

gkellogg committed:
First pass at ContextResolver.

gkellogg committed:
Fixes to _resolve\_context\_urls_.

gkellogg committed:
Record `@context: null` as `@context: false`in a scoped context  so that it doesn't get dropped when handling scoped contexts

gkellogg committed:
Attempt some module initialization for ContextResolver and Canonicalization.

gkellogg committed:
Have jsonld export ContextResolver.

gkellogg committed:
Remove some attempts at module initialization which didn't help :(.

gkellogg committed:
Add initialization to c14n/__init__.py.

gkellogg committed:
Add some initialization for ContextResolver to pyld/__init__.py

gkellogg committed:
Fix EARL output to not use ".jsonld" suffix to test manifests.

gkellogg committed:
Only cache a resolve context for Python >= 3.6, due to intermittent runtime issues.

gkellogg committed:
Use cache for imported document as well.

(Note minor potential for colliding on the ResolvedContext cache).

gkellogg committed:
Use a separate LRU cache for inverse contexts to avoid side-effect modifications to the active context.

gkellogg committed:
_process_context returns a frozen context.

gkellogg committed:
Doesn't work with Python < 3.6.

gkellogg committed:
* Don't set `@base` in initial context and don't resolve a relative IRI
    when setting `@base` in a context, so that the document location can
    be kept separate from the context itself.
  * Pass in document location to `_expand_iri` (through 'base' option) and use when resolving document-relative IRIs.
  * Pass in document location to `_compact_iri` (through 'base' option) and use when compacting
    document-relative IRIs.

gkellogg committed:
Use static initial contexts composed of just `mappings` and `processingMode` to enhance preprocessed context cachability.

gkellogg committed:
Use UUIDs in active context for cache key for resolved context.

gkellogg committed:
Update more pending test patterns.

gkellogg committed:
Add frozendict and uuid to setup.

gkellogg committed:
Don't deep freeze context objects, just freeze the top-level dict.
Use the canonicalization of a local context as the ContextResolver key
for a local context, rather than it's object identity (`id()`),
which is not guaranteed to be the same for each use.

davidlehn committed:
Improve markup.

Escape `_` related terms as needed.

davidlehn committed:
Word wrap.

davidlehn committed:
Switch to argparse and improve test running.

* Switched to argparse.
* **BREAKING**: Removed `-d` and `-m` test runner options in favor of
  just listing as arguments.
* If no test manifests or directories are specified, default to sibling
  directories for json-ld-api, json-ld-framing, and normalization.

davidlehn committed:
Use new testing command syntax.

gkellogg commented:
@davidlehn thanks for updating the test runner, much easier now.

gkellogg committed:
Ignore profiler directory.

gkellogg committed:
Set ONLY_IDENTIFIER earlier in test suite.

(cherry picked from commit 92fa1798aad5332d6063e6f43e6a9bfd4c61fc6c)

davidlehn committed:
Quote function name.

davidlehn committed:
Remove trailing whitespace.

davidlehn committed:
Update to https URLs.
